### DISCUSSION TITLE
Enhance autocomplete/tooltip of function which uses `ParamSpec` and the signature ends with `*args: P.args, **kwargs: P.kwargs`

### DISCUSSION THREAD

**Discussion by michaeloliverx** [microsoft/pylance-release#2634] [Enhancement]
Take the following snippet:

```python
from collections.abc import Callable
from typing import TypeVar

from typing_extensions import ParamSpec

T = TypeVar("T")
P = ParamSpec("P")


async def run_in_threadpool(func: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    ...


def foo(bar: str) -> bool:
    ...


async def main():
    result = await run_in_threadpool(foo, bar=")
```

When I write `run_in_threadpool(foo, ` the tooltip shows me the following:

![image](https://user-images.githubusercontent.com/55017335/157013742-8c1064ce-5f1e-4363-9bb4-1c559f64f7ec.png)

I think it would be a better user experience if it was able to show me args and kwargs of `foo`.

**Comment by erictraut** [comment]
What you're suggesting is that pylance "solve the type variables and param specs" for a partially-completed call. I think that might be viable in the very specific case that you've highlighted in your example, but it will lead to misleading (and incorrect) signatures in the more general case. So if we were to add this, it would need to be limited specifically to the case for ParamSpecs where the signature ends with `*args: P.args, **kwargs: P.kwargs` and all of the previous arguments have already been provided. That seems very niche to me, but we could consider adding this if we receive enough signal from other users — in the form of additional upvotes.

**Comment by michaeloliverx** [comment]
I think this pattern is somewhat common:
- [`functools.partial`](https://docs.python.org/3/library/functools.html#functools.partial)
- [`starlette.concurrency.run_in_threadpool`](https://github.com/encode/starlette/blob/6798311a9154300cfea53f8a99c454807c9690ed/starlette/concurrency.py#L35-L37)

---

I have come across this call pattern in TypeScript quite a bit and its trivial to type it and autocomplete/tooltip works fine:

```typescript
function registerCallback<Fn extends (...args: any) => any>(func: Fn, ...args: Parameters<Fn>): void {
    //
}

function foo(options: {foo: string, nums: Array<number>}): boolean {
    return true;
}
```


![image](https://user-images.githubusercontent.com/55017335/157267055-8b5ca938-1553-47d8-98e4-700cb5441d65.png)

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by jakkes** [comment]
+1 would like this.
